main.go

```diff
@@ -0,0 +1,182 @@
+	host := flag.String("host", " ", "Host Name or IP Address")
+	cakey := flag.String("cakey", " ", "Private Key Certificate Chain")
+	cacert := flag.String("cacert", " ", "Certificate Authority")
+	concurrency := flag.Int("concurrency", 0, "concurrency")
+	sleep := flag.Duration("sleep", 0, "sleep in seconds")
+	duration := flag.Int("duration", 0, "Amount of files to be synced to the specified zone")
+	flag.Parse()
+
+	rootCAs, err := ioutil.ReadFile(*cacert)
+	if err != nil {
+		log.Fatalf("CLIENT: Error while loading CA files %s", err.Error())
+	}
+
+	key, err := ioutil.ReadFile(*cakey)
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to load Certificate Authority private key: %s", err.Error())
+	}
+
+	decode, _ := pem.Decode(key)
+	if decode == nil {
+		log.Fatalf("CLIENT: Failed to parse Certificate Authority private key")
+	}
+
+	parsedKey, err := x509.ParsePKCS1PrivateKey(decode.Bytes)
+	if err != nil {
+		log.Fatalf("CLIENT: Unable to parse RSA private key: %s", err)
+	}
+
+	privateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to generate new private key %s", err)
+	}
+
+	privatePem := new(bytes.Buffer)
+	 pem.Encode(privatePem, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	})
+
+	block, _ := pem.Decode(rootCAs)
+	if block == nil {
+		log.Fatalf("CLIENT: Failed to parse certificate PEM")
+	}
+	CAuthority, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to parse certificate: %s", err.Error())
+	}
+
+	tlsConfigs := make(map[int]*tls.Config)
+
+	for i := 0; i < *concurrency; i++ {
+		_ , certPem := GenerateCertificate(CAuthority, parsedKey, privateKey, duration, fmt.Sprintf("localhost%d", i))
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(rootCAs)
+
+		cert, err := tls.X509KeyPair(certPem.Bytes(), privatePem.Bytes())
+		if err != nil {
+			log.Fatalf("CLIENT: Failed to load certificate and private key: %s", err.Error())
+		}
+
+		conf := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth: tls.RequireAnyClientCert,
+			ClientCAs: caCertPool,
+			InsecureSkipVerify: true,
+		}
+
+		conf.Rand = rand.Reader
+		tlsConfigs[i] = conf
+	}
+
+	for i := 0; i < *concurrency; i++ {
+		go ProcessOutgoingConnection(host, port, tlsConfigs[i], sleep)
+	}
+
+	fmt.Scanln()
+}
+
+func ProcessOutgoingConnection(host *string, port *int, conf *tls.Config, sleep *time.Duration)  {
+	Addr := fmt.Sprintf("%s%s%d", *host, ":", *port)
+
+	conn, err := tls.Dial("tcp", Addr, conf)
+
+	defer conn.Close()
+	if err != nil {
+		log.Fatalf("CLIENT: Failed to connect: %s", err.Error())
+	}
+
+	log.Printf("CLIENT: Connect to %s succeed", Addr)
+	time.Sleep(*sleep)
+
+	encoder := &Encoder{
+		JSONRPC: "2.0",
+		Method: "pki::RequestCertificate",
+		Params: map[string]string{
+			"ticket": " ",
+		},
+	}
+
+	message, err := json.Marshal(&encoder)
```

Comment:
... and this only once for all.

Reply:
I do not understand what exactly you mean!

Reply:
Addr and message are equal for all (let’s say) 100 connections, right?

Reply:
yeah exactly.

Reply:
And why do you compute the very same values 100 times?

Reply:
Do you want me to encode the message in the main function and pass it as a parameter to the ``ProcessOutgoingConnection()``  function?

Reply:
... and the same for Addr.